options/posix/generic/sys-uio.cpp

```diff
@@ -13,14 +18,32 @@ ssize_t writev(int fd, const struct iovec *iovs, int iovc) {
 	__ensure(iovc);
 
 	ssize_t written = 0;
+	size_t bytes = 0;
 	for(int i = 0; i < iovc; i++) {
-		__ensure(iovs[i].iov_len);
-		auto chunk = write(fd, iovs[i].iov_base, iovs[i].iov_len);
-		// TODO: writev() should be atomic.
-		// This emulation here is not correct if there is an error!
-		__ensure(chunk > 0);
-		written += chunk;
+		if(SSIZE_MAX - bytes < iovs[i].iov_len) {
+			errno = EINVAL;
+			return -1;
+		}
+		bytes += iovs[i].iov_len;
 	}
+	char *buffer = (char *)malloc(bytes);
```

Comment:
We can use a `frg::vector<char, MemoryAllocator>` instead of `malloc`. (Use `.resize()` to resize the buffer and `.data()` to access the `char *`.)